SPECS/kernel-hci/kernel-hci.spec

```diff
@@ -61,6 +61,7 @@ Requires:       filesystem
 Requires:       kmod
 Requires(post): coreutils
 Requires(postun): coreutils
+Provides: kernel = %{version}-%{release}
```

Comment:
I'm not sure, if we can do that, since this will conflict with the default `kernel` package. Especially with the `Release` bump in this one. If you try `tdnf upgrade kernel` on a system running the default `kernel` package, I think you would end up upgrading to `kernel-hci` and that's probably not what you want.

Reply:
Might need to do some research/experiments here. There is precedent for multiple packages providing the same thing (busybox, etc). I'm not 100% how tdnf will handle this though. The obvious solution is to leave the package you already have installed... but I'd not put money on that being the actual case.

Reply:
Did some tests, unfortunately `tdnf` does not work well here. `dnf` works great though.

> I have two packages, `pkg_a`, and `pkg_b`. Both of these have `Provide: test_pkg`. They have a Conflicts: with each other.
`tdnf -y install test_pkg && tdnf -y erase test_pkg` adds and removes A.
`tdnf -y install pkg_b` installs B as expected.
`tdnf -y install test_pkg` accurately calculates the "capability match" for `test_pkg`, but it then picks the first valid option to provide it (`pkg_a`). Unfortunately, these conflict with each other and tdnf fails.
`dnf -y install test_pkg` notices that it already has an implementation of `test_pkg` and does nothing.

Reply:
I would be curious to see how microdnf handles this use case, may be another reason to take a look at it.

Reply:
Turns out we have microdnf in extended, seems to handle this use case reasonably well.

Reply:
@PawelWMS On my system with kernel 5.15.77.1-2 installed and kernel 5.15.79.1-1 and kernel-hci 5.15.79.1-2 available in repos, tdnf behaves appropriately:

```
root [ /home/viperri ]# tdnf upgrade
Loaded plugin: tdnfrepogpgcheck

Upgrading:
sudo                                                 x86_64                     1.9.12p1-1.cm2                     mariner-official-base          7.44M 7805468
mariner-release                                      noarch                     2.0-27.cm2                         mariner-official-base            563.00b 563
kernel-devel                                         x86_64                     5.15.79.1-1.cm2                    mariner-official-base        49.18M 51571281
kernel-tools                                         x86_64                     5.15.79.1-1.cm2                    mariner-official-base        45.14M 47330896
kernel                                               x86_64                     5.15.79.1-1.cm2                    mariner-official-base        40.99M 42982257
hyperv-daemons-license                               noarch                     5.15.79.1-1.cm2                    mariner-official-base            496.00b 496
hypervvssd                                           x86_64                     5.15.79.1-1.cm2                    mariner-official-base           19.70k 20177
hypervkvpd                                           x86_64                     5.15.79.1-1.cm2                    mariner-official-base           38.25k 39173
hypervfcopyd                                         x86_64                     5.15.79.1-1.cm2                    mariner-official-base           15.68k 16059
hyperv-daemons                                       x86_64                     5.15.79.1-1.cm2                    mariner-official-base                0.00b 0

Total installed size: 142.83M 149766370
Is this ok [y/N]: N
Error(1032) : Operation aborted.
root [ /home/viperri ]# tdnf list | grep "^kernel.x86_64"
kernel.x86_64                               5.15.77.1-2.cm2          @System
kernel.x86_64                               5.15.32.1-3.cm2     mariner-official-base
kernel.x86_64                               5.15.41.1-1.cm2     mariner-official-base
kernel.x86_64                               5.15.45.1-2.cm2     mariner-official-base
kernel.x86_64                               5.15.48.1-2.cm2     mariner-official-base
kernel.x86_64                               5.15.48.1-4.cm2     mariner-official-base
kernel.x86_64                               5.15.55.1-1.cm2     mariner-official-base
kernel.x86_64                               5.15.57.1-1.cm2     mariner-official-base
kernel.x86_64                               5.15.57.1-3.cm2     mariner-official-base
kernel.x86_64                               5.15.67.1-4.cm2     mariner-official-base
kernel.x86_64                               5.15.70.1-1.cm2     mariner-official-base
kernel.x86_64                               5.15.74.1-3.cm2     mariner-official-base
kernel.x86_64                               5.15.77.1-1.cm2     mariner-official-base
kernel.x86_64                               5.15.79.1-1.cm2     mariner-official-base
root [ /home/viperri ]# tdnf list | grep "^kernel-hci.x86_64"
kernel-hci.x86_64                           5.15.79.1-2.cm2       built_rpms
kernel-hci.x86_64                           5.15.63.1-1.cm2     mariner-official-base
kernel-hci.x86_64                           5.15.70.1-1.cm2     mariner-official-base
kernel-hci.x86_64                           5.15.74.1-1.cm2     mariner-official-base
kernel-hci.x86_64                           5.15.77.1-1.cm2     mariner-official-base
kernel-hci.x86_64                           5.15.79.1-1.cm2     mariner-official-base
```

@dmcilvaney I don't know why tdnf doesn't behave as expected in your tests, but wouldn't the test be closer to what this PR is trying to do if pkg_a provided no additional capabilities, and pkg_b provided pkg_a? That would be more in line with kernel-hci providing kernel, which is what this PR adds.

I rebuilt the packages with the following capabilities:

```
# rpm -qp --provides /usr/src/mariner/RPMS/pkg_a-1.0-2.x86_64.rpm
pkg_a = 1.0-2
pkg_a(x86-64) = 1.0-2
# rpm -qp --provides /usr/src/mariner/RPMS/pkg_b-1.0-2.x86_64.rpm
pkg_a = 1.0-2
pkg_b = 1.0-2
pkg_b(x86-64) = 1.0-2
```

The following tests select the right package and behave as expected:

* `tdnf install pkg_a && tdnf erase pkg_a` (pkg_a installed then erased)
* `tdnf install pkg_b && tdnf erase pkg_b` (pkg_b installed then erased)
* `tdnf install pkg_b && tdnf erase pkg_a` (pkg_b installed then not erased)
* `tdnf install pkg_a && tdnf erase pkg_b` (pkg_a installed then not erased)

Reply:
The kernel package will also take precedence in tdnf's dependency resolution. Here I added a `pkg_c` that requires capability `pkg_a` and declares no additional capabilities of its own. When a user just wants to install `pkg_c`, `pkg_a` will also be installed. However, if the user explicitly installs `pkg_b` in addition to `pkg_c`, `pkg_a` will be properly ignored:

```
root [ /home/viperri ]# tdnf install pkg_c
Loaded plugin: tdnfrepogpgcheck

Installing:
pkg_a                                                x86_64                     1.0-2                              built_rpms                           6.00b 6
pkg_c                                                x86_64                     1.0-2                              built_rpms                           6.00b 6

Total installed size:  12.00b 12
Is this ok [y/N]: N
Error(1032) : Operation aborted.
root [ /home/viperri ]# tdnf install pkg_b pkg_c
Loaded plugin: tdnfrepogpgcheck

Installing:
pkg_c                                                x86_64                     1.0-2                              built_rpms                           6.00b 6
pkg_b                                                x86_64                     1.0-2                              built_rpms                           6.00b 6

Total installed size:  12.00b 12
Is this ok [y/N]: N
Error(1032) : Operation aborted.
```

This corresponds to something like `tdnf install alsa-utils` (which will install kernel-drivers-sound) vs `tdnf install alsa-utils kernel-hci-drivers-sounds` (which won't).

Reply:
`tdnf install pkg_b && tdnf install pkg_c` will install `pkg_b` then just `pkg_c` since `pkg_b` is already installed and provides the capability `pkg_a`. Also, when you do a `tdnf remove pkg_b` after installing `pkg_b` and `pkg_c`, `pkg_a` will want to replace `pkg_b` with `pkg_a` since `pkg_c` still requires the capability.

```
root [ /home/viperri ]# tdnf install pkg_b
Loaded plugin: tdnfrepogpgcheck

Installing:
pkg_b                                                x86_64                     1.0-2                              built_rpms                           6.00b 6

Total installed size:   6.00b 6
Is this ok [y/N]: y

Downloading:
pkg_b                                     6823   100%
Testing transaction
Running transaction
Installing/Updating: pkg_b-1.0-2.x86_64

Complete!
root [ /home/viperri ]# tdnf install pkg_c
Loaded plugin: tdnfrepogpgcheck

Installing:
pkg_c                                                x86_64                     1.0-2                              built_rpms                           6.00b 6

Total installed size:   6.00b 6
Is this ok [y/N]: y

Downloading:
pkg_c                                     6825   100%
Testing transaction
Running transaction
Installing/Updating: pkg_c-1.0-2.x86_64

Complete!
root [ /home/viperri ]# tdnf remove pkg_b
Loaded plugin: tdnfrepogpgcheck

Installing:
pkg_a                                                x86_64                     1.0-2                              built_rpms                           6.00b 6

Total installed size:   6.00b 6

Removing:
pkg_b                                                x86_64                     1.0-2                              @System                              6.00b 6

Total installed size:   6.00b 6
Is this ok [y/N]:
```